lib/datasets/mnist.rb

```diff
@@ -5,6 +5,7 @@
 module Datasets
   class MNIST < Dataset
     BASE_URL = "http://yann.lecun.com/exdb/mnist/"
+    FALLBACK_URL = "https://ossci-datasets.s3.amazonaws.com/mnist/"
```

Comment:
This may be used by subclasses. How about introducing `#base_urls` instead?

```diff
diff --git a/lib/datasets/fashion-mnist.rb b/lib/datasets/fashion-mnist.rb
index 9e5bcf6..194758f 100644
--- a/lib/datasets/fashion-mnist.rb
+++ b/lib/datasets/fashion-mnist.rb
@@ -2,9 +2,13 @@ require_relative 'mnist'
 
 module Datasets
   class FashionMNIST < MNIST
-    BASE_URL = "http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/"
-
     private
+    def base_urls
+      [
+        "http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/",
+      ]
+    end
+
     def dataset_name
       "Fashion-MNIST"
     end
diff --git a/lib/datasets/kuzushiji-mnist.rb b/lib/datasets/kuzushiji-mnist.rb
index 344bf06..9aa69ff 100644
--- a/lib/datasets/kuzushiji-mnist.rb
+++ b/lib/datasets/kuzushiji-mnist.rb
@@ -2,9 +2,13 @@ require_relative 'mnist'
 
 module Datasets
   class KuzushijiMNIST < MNIST
-    BASE_URL = "http://codh.rois.ac.jp/kmnist/dataset/kmnist/"
-
     private
+    def base_urls
+      [
+        "http://codh.rois.ac.jp/kmnist/dataset/kmnist/",
+      ]
+    end
+
     def dataset_name
       "Kuzushiji-MNIST"
     end
diff --git a/lib/datasets/mnist.rb b/lib/datasets/mnist.rb
index b07aa4e..9185019 100644
--- a/lib/datasets/mnist.rb
+++ b/lib/datasets/mnist.rb
@@ -4,8 +4,6 @@ require_relative "dataset"
 
 module Datasets
   class MNIST < Dataset
-    BASE_URL = "http://yann.lecun.com/exdb/mnist/"
-
     class Record < Struct.new(:data, :label)
       def pixels
         data.unpack("C*")
@@ -27,7 +25,7 @@ module Datasets
 
       @metadata.id = "#{dataset_name.downcase}-#{type}"
       @metadata.name = "#{dataset_name}: #{type}"
-      @metadata.url = self.class::BASE_URL
+      @metadata.url = base_urls.first
       @metadata.licenses = licenses
       @type = type
 
@@ -44,7 +42,7 @@ module Datasets
 
       image_path = cache_dir_path + target_file(:image)
       label_path = cache_dir_path + target_file(:label)
-      base_url = self.class::BASE_URL
+      base_url = base_urls.first
 
       download(image_path, base_url + target_file(:image))
       download(label_path, base_url + target_file(:label))
@@ -53,6 +51,13 @@ module Datasets
     end
 
     private
+    def base_urls
+      [
+        "http://yann.lecun.com/exdb/mnist/",
+        "https://ossci-datasets.s3.amazonaws.com/mnist/",
+      ]
+    end
+
     def licenses
       []
     end
```

Reply:
I missed used by subclasses. Thanks! I fixed it.